subsys/bluetooth/controller/hci/hci.c

```diff
@@ -4798,6 +4800,147 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+	cpu_data->a4 = sys_cpu_to_le32(esf->basic.a4);
+	cpu_data->ip = sys_cpu_to_le32(esf->basic.ip);
+	cpu_data->lr = sys_cpu_to_le32(esf->basic.lr);
+	cpu_data->xpsr = sys_cpu_to_le32(esf->basic.xpsr);
+}
+#endif /* CONFIG_CPU_CORTEX_M */
+
+static struct net_buf *vs_err_evt_create(uint8_t subevt, uint8_t len)
+{
+	struct net_buf *buf;
+
+	buf = net_buf_alloc(&vs_err_tx_pool, K_FOREVER);
+	if (buf) {
+		struct bt_hci_evt_le_meta_event *me;
+		struct bt_hci_evt_hdr *hdr;
+
+		net_buf_reserve(buf, BT_BUF_RESERVE);
+		bt_buf_set_type(buf, BT_BUF_EVT);
+
+		hdr = net_buf_add(buf, sizeof(*hdr));
+		hdr->evt = BT_HCI_EVT_VENDOR;
+		hdr->len = len + sizeof(*me);
+
+		me = net_buf_add(buf, sizeof(*me));
+		me->subevent = subevt;
+	}
+
+	return buf;
+}
+
+struct net_buf *hci_vs_err_stack_frame(unsigned int reason, const z_arch_esf_t *esf)
+{
+	/* Prepare vendor specific HCI Fatal Error event */
+	struct bt_hci_vs_fatal_error_stack_frame *sf;
+	bt_hci_vs_fatal_error_cpu_data *cpu_data;
+	struct net_buf *buf;
+
+	buf = vs_err_evt_create(BT_HCI_EVT_VS_ERROR_DATA_TYPE_STACK_FRAME,
+				sizeof(*sf) + sizeof(*cpu_data));
+	if(buf != NULL) {
+		sf = net_buf_add(buf, (sizeof(*sf) + sizeof(*cpu_data)));
+		sf->reason = sys_cpu_to_le32(reason);
+		sf->cpu_type = BT_HCI_EVT_VS_ERROR_CPU_TYPE_CORTEX_M;
+
+		vs_err_fatal_cpu_data_fill(
+			(bt_hci_vs_fatal_error_cpu_data *)sf->cpu_data, esf);
+	} else {
+		BT_ERR("Can't create HCI Fatal Error event");
+
+	}
+
+	return buf;
+}
+
+static struct net_buf *hci_vs_err_trace_create(uint8_t data_type,
+					       const char *file_path,
+					       uint32_t line, uint64_t pc)
+{
+	uint32_t len = 0U, pos = 0U;
+	struct net_buf *buf = NULL;
+
+	if (file_path) {
+		/* Extract file name from a path */
+		while (file_path[len] != '\0') {
+			if (file_path[len] == '/') {
+				pos = len + 1;
+			}
+			len++;
+		}
+		file_path += pos;
+		len -= pos;
+
+		/* If file name was found in file_path, in other words: file_path is not empty
+		 * string and is not `foo/bar/`.
+		 */
+		if (len) {
+			/* Total data length: len = file name strlen + \0 + sizeof(line number) */
+			uint32_t data_len = len + 1 + sizeof(line);
+
+			/* If a buffer is created for a TRACE data, include sizeof(pc) in total
+			 * length.
+			 */
+			if (data_type == BT_HCI_EVT_VS_ERROR_DATA_TYPE_TRACE) {
+				data_len += sizeof(pc);
+			}
+
+			/* Maximum length of an HCI event data is UINT8_MAX */
+			if (data_len < UINT8_MAX) {
+				/* Prepare vendor specific HCI Fatal Error event */
+				buf = vs_err_evt_create(data_type, data_len);
+				if(buf != NULL) {
+					if (data_type == BT_HCI_EVT_VS_ERROR_DATA_TYPE_TRACE) {
+						net_buf_add_le64(buf, pc);
+					}
+					net_buf_add_mem(buf, file_path, len);
+					net_buf_add_u8(buf, STR_NULL_TERMINATOR);
+					net_buf_add_le32(buf, line);
+				} else {
+					BT_ERR("Can't create HCI Fatal Error event");
+				}
+			} else {
+				BT_ERR("Too much data for a HCI Fatal Error event");
+			}
```

Comment:
Could we simply truncate the filename? Or does it need to be complete for the event to properly work?

Reply:
Thank for a good tip, the filename may be truncated. I thing that file name with length around 200 characters is unique enough that it can be truncated. 

BTW the actual max length should be `BT_BUF_RX_SIZE` not `UINT8_MAX`. I'll use the `BT_BUF_RX_SIZE` in the code.